File: C99/백준/Gold/1938. 통나무 옮기기/통나무 옮기기.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <stdint.h>

/**
 * @description BFS 시뮬레이션
 * @performance 메모리: 1,112 KB, 동작시간: 0 ms
 * @author java08
 */

// 현재 상태에서 각 움직임을 수행할 수 있는지 확인하고, 큐에 넣는다.
// visited[N][N][2]를 통해 해당 좌표에서 가로/세로로 놓인 적 있는지 검사해서 가지치기를 한다.

int N, bCnt, eCnt;
int **map;
bool ***dp;
int b[3][2];
int e[3][2];
int B[3]; // r, c, d
int E[3];

typedef struct
{
    int r, c, d;
} pos;

bool up(int r, int c, int d)
{
    if (d == 0)
    {                  // 가로 방향
        if (r - 1 < 0) // 중심점을 기준으로 바로 윗 행만 고려하면 됨
            return false;
        if (map[r - 1][c - 1] != 0 || map[r - 1][c] != 0 || map[r - 1][c + 1] != 0)
            return false;
        return true;
    }
    else
    {                  // 세로 방향
        if (r - 2 < 0) // 중심점을 기준으로 2칸 위를 고려해야 함
            return false;
        if (map[r - 2][c] != 0)
            return false;
        return true;
    }
}

bool down(int r, int c, int d)
{
    if (d == 0)
    {                   // 가로 방향
        if (r + 1 >= N) // 중심점을 기준으로 바로 아래 행만 고려하면 됨
            return false;
        if (map[r + 1][c - 1] != 0 || map[r + 1][c] != 0 || map[r + 1][c + 1] != 0)
            return false;
        return true;
    }
    else
    {                   // 세로 방향
        if (r + 2 >= N) // 중심점을 기준으로 2칸 아래를 고려해야 함
            return false;
        if (map[r + 2][c] != 0)
            return false;
        return true;
    }
}

bool right(int r, int c, int d)
{
    if (d == 0)
    {                   // 가로 방향
        if (c + 2 >= N) // 중심점을 기준으로 2칸 오른쪽을 고려해야 함
            return false;
        if (map[r][c + 2] != 0)
            return false;
        return true;
    }
    else
    {                   // 세로 방향
        if (c + 1 >= N) // 중심점을 기준으로 1칸 오른쪽 고려해야 함
            return false;
        if (map[r - 1][c + 1] != 0 || map[r][c + 1] != 0 || map[r + 1][c + 1] != 0)
            return false;
        return true;
    }
}

bool left(int r, int c, int d)
{
    if (d == 0)
    {                  // 가로 방향
        if (c - 2 < 0) // 중심점을 기준으로 2칸 왼쪽을 고려해야 함
            return false;
        if (map[r][c - 2] != 0)
            return false;
        return true;
    }
    else
    {                  // 세로 방향
        if (c - 1 < 0) // 중심점을 기준으로 1칸 왼쪽 고려해야 함
            return false;
        if (map[r - 1][c - 1] != 0 || map[r][c - 1] != 0 || map[r + 1][c - 1] != 0)
            return false;
        return true;
    }
}

bool turn(int r, int c, int d)
{
    if (r - 1 < 0 || c - 1 < 0 || r + 1 >= N || c + 1 >= N)
        return false;
    for (int i = r - 1; i <= r + 1; i++)
    {
        for (int j = c - 1; j <= c + 1; j++)
        {
            if (map[i][j] != 0)
                return false;
        }
    }
    return true;
}

int main(void)
{
    // 입력 시작
    if (scanf("%d", &N) != 1)
        return 1;
    map = malloc(N * sizeof(int *));
    dp = malloc(N * sizeof(bool **));
    for (int i = 0; i < N; i++)
    {
        map[i] = malloc(N * sizeof(int));
        dp[i] = malloc(N * sizeof(bool *));
        for (int j = 0; j < N; j++)
        {
            dp[i][j] = calloc(2, sizeof(bool));
        }
    }

    char *in = malloc((N + 1) * sizeof(char));
    for (int i = 0; i < N; i++)
    {
        if (scanf("%s", in) != 1)
            return 1;
        for (int j = 0; j < N; j++)
        {
            if (in[j] == 'B')
            {
                b[bCnt][0] = i;
                b[bCnt][1] = j;
                bCnt++;
                map[i][j] = 0;
            }
            else if (in[j] == 'E')
            {
                e[eCnt][0] = i;
                e[eCnt][1] = j;
                eCnt++;
                map[i][j] = 0;
            }
            else
            {
                map[i][j] = in[j] - '0';
            }
        }
    }
    free(in);
    // 입력 끝

    B[0] = b[1][0];
    B[1] = b[1][1];
    if (b[0][0] == b[1][0]) // 행이 같으면 가로방향
    {
        B[2] = 0;
    }
    else
    {
        B[2] = 1;
    }

    E[0] = e[1][0];
    E[1] = e[1][1];
    if (e[0][0] == e[1][0]) // 행이 같으면 가로방향
    {
        E[2] = 0;
    }
    else
    {
        E[2] = 1;
    }

    pos q[10000];
    int front = 0, rear = 0;
    pos init = {B[0], B[1], B[2]};
    q[rear++] = init;
    dp[init.r][init.c][init.d] = true;
    int result = 0;
    bool flag = false;
    while (front < rear)
    {
        int cnt = rear - front;
        while (cnt--)
        {
            pos cur = q[front++];
            int curR = cur.r;
            int curC = cur.c;
            int curD = cur.d;
            if (curR == E[0] && curC == E[1] && curD == E[2])
            {
                flag = true;
                break;
            }
            int nextR, nextC, nextD;
            if (up(curR, curC, curD))
            {
                nextR = curR - 1;
                nextC = curC;
                nextD = curD;
                if (!dp[nextR][nextC][nextD])
                {
                    dp[nextR][nextC][nextD] = true;
                    pos next = {nextR, nextC, nextD};
                    q[rear++] = next;
                }
            }
            if (down(curR, curC, curD))
            {
                nextR = curR + 1;
                nextC = curC;
                nextD = curD;
                if (!dp[nextR][nextC][nextD])
                {
                    dp[nextR][nextC][nextD] = true;
                    pos next = {nextR, nextC, nextD};
                    q[rear++] = next;
                }
            }
            if (right(curR, curC, curD))
            {
                nextR = curR;
                nextC = curC + 1;
                nextD = curD;
                if (!dp[nextR][nextC][nextD])
                {
                    dp[nextR][nextC][nextD] = true;
                    pos next = {nextR, nextC, nextD};
                    q[rear++] = next;
                }
            }
            if (left(curR, curC, curD))
            {
                nextR = curR;
                nextC = curC - 1;
                nextD = curD;
                if (!dp[nextR][nextC][nextD])
                {
                    dp[nextR][nextC][nextD] = true;
                    pos next = {nextR, nextC, nextD};
                    q[rear++] = next;
                }
            }
            if (turn(curR, curC, curD))
            {
                nextR = curR;
                nextC = curC;
                nextD = curD ? 0 : 1;
                if (!dp[nextR][nextC][nextD])
                {
                    dp[nextR][nextC][nextD] = true;
                    pos next = {nextR, nextC, nextD};
                    q[rear++] = next;
                }
            }
        }
        if (flag)
            break;
        result++;
    }
    if (!flag)
        result = 0;
    printf("%d\n", result);
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            free(dp[i][j]);
        }
        free(dp[i]);
        free(map[i]);
    }
    free(dp);
    free(map);
    return 0;
}

File: Java/백준/Gold/1938. 통나무 옮기기/통나무 옮기기.java
import java.io.*;
import java.util.*;

/**
 * @description BFS + 시뮬레이션
 * @performance 메모리: ? KB, 동작시간: ? ms
 * @author python98
 */
public class Main {

	private static final int[][] DIR = {{-1,0},{1,0},{0,-1},{0,1}}; // 위 아래 좌 우
	private static final int[][][] HORI_VERT = {{{0,-1},{0,0},{0,1}},{{-1,0},{0,0},{1,0}}};
	private static int N;
	private static long[] map; // 64bit N개로 비트마스킹
	private static long[][] dp; // 0:가로 1:세로
	private static int[] B, E;

	static class Pos{
		int r, c, d;
		Pos(int r, int c, int d){
			this.r = r;
			this.c = c;
			this.d = d;
		}
	}

	private static boolean isPossible(Pos cur, int idx){
		if(idx==4) { // 회전
			if(cur.r-1<0||cur.r+1>=N||cur.c-1<0||cur.c+1>=N) return false;
			for(int i = cur.r-1 ; i <= cur.r+1 ; i++) {
				for(int j = cur.c-1 ; j <= cur.c+1 ; j++) {
					if(((map[i]>>j)&1)==1) return false;
				}
			}
			return true;
		}else {
			for(int i = 0 ; i < 3 ; i++){
				int nr = cur.r + HORI_VERT[cur.d][i][0] + DIR[idx][0]; // 막대 3개 중 위치 + 이동 방향
				int nc = cur.c + HORI_VERT[cur.d][i][1] + DIR[idx][1];
				if(nr<0||nc<0||nr>=N||nc>=N) return false;
				if(((map[nr]>>nc)&1)==1) return false;
			}
			return true;
		}
	}
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		N = Integer.parseInt(br.readLine());
		map = new long[N];
		dp = new long[2][N];

		int[][] b = new int[3][2];
		int[][] e = new int[3][2];
		int bCnt = 0, eCnt = 0;
		for(int i = 0 ; i < N ; i++) {
			char[] input = br.readLine().toCharArray();
			for(int j = 0 ; j < N ; j++) {
				long bit = 0l;
				if(input[j]=='B') {
					b[bCnt][0] = i;
					b[bCnt][1] = j;
					bCnt++;
				}else if(input[j]=='E') {
					e[eCnt][0] = i;
					e[eCnt][1] = j;
					eCnt++;
				}else {
					bit = input[j]-'0';
				}
				map[i]|=(bit<<j);
			}
		}
		B = new int[3];
		B[0] = b[1][0];
		B[1] = b[1][1];
		B[2] = b[0][0]==b[1][0] ? 0:1;
		E = new int[3];
		E[0] = e[1][0];
		E[1] = e[1][1];
		E[2] = e[0][0]==e[1][0] ? 0:1;

		Pos[] q = new Pos[10000];
		int front = 0, rear = 0;
		q[rear++] = new Pos(B[0], B[1], B[2]);
		dp[B[2]][B[0]]|=(1l<<B[1]);
		boolean flag = false;
		int result = 0;
		while(front<rear){
			int cnt = rear-front;
			while(cnt-->0){
				Pos cur = q[front++];
				if(cur.r==E[0]&&cur.c==E[1]&&cur.d==E[2]) {
					flag = true;
					break;
				}
				for(int i = 0 ; i < 5 ; i++) {
					if(isPossible(cur,i)){
						int nextR, nextC, nextD;
						if(i==4) {
							nextR = cur.r;
							nextC = cur.c;
							nextD = cur.d==0?1:0;
						} else {
							nextR = cur.r+DIR[i][0];
							nextC = cur.c+DIR[i][1];
							nextD = cur.d;
						}
						if(((dp[nextD][nextR]>>nextC)&1)==0) {
							dp[nextD][nextR]|=(1l<<nextC);
							q[rear++] = new Pos(nextR, nextC, nextD);
						}
					}
				}
			}
			if(flag){
				break;
			}
			result++;
		}
		if(!flag) result = 0;
		System.out.print(result);
	}
}